roadmap/implementers-guide/src/node/availability/availability-recovery.md

```diff
@@ -20,3 +22,148 @@ Output:
+
+#### `FromInteraction::Concluded`
+
+1. Load the entry from the `interactions` map. It should always exist, if not for logic errors. Send the result to each member of `awaiting`.
+1. Add the entry to the availability_lru.
+
+#### `FromInteraction::MakeRequest(discovery_pub, candidate_hash, validator_index, response)`
+
+1. Add an `AwaitedRequest` to the `discovering_validators` map under `discovery_pub`.
+1. Issue a `NetworkBridgeMessage::ConnectToValidators`.
+1. Add the stream of connected validator events to `state.connecting_validators`.
+
+#### `FromInteraction::ReportPeer(peer, rep)`
+
+1. Issue a `NetworkBridgeMessage::ReportPeer(peer, rep).
+
+### Responding to network events.
+
+#### On `connecting_validators` event:
+
+1. If the validator exists under `discovering_validators`, remove the entry.
+1. For each `AwaitedChunk` in the entry, issue a `AvailabilityRecoveryV1Message::RequestChunk(next_request_id, candidate_hash, validator_index)` and make an entry in the `live_chunk_requests` map.
+
+#### On receiving `AvailabilityRecoveryV1::RequestChunk(r_id, candidate_hash, validator_index)`
+
+1. Issue a `AvailabilityStore::QueryChunk(candidate-hash, validator_index, response)` message.
+1. Whatever the result, issue a `AvailabilityRecoveryV1Message::Chunk(r_id, response)` message.
+
+#### On receiving `AvailabilityRecoveryV1::Chunk(r_id, chunk)`
+
+1. If there exists an entry under `r_id`, remove it. If there doesn't exist one, report the peer and return. If the peer in the entry doesn't match the sending peer, reinstate the entry, report the peer, and return.
+1. Send the chunk response on the `awaited_chunk` for the interaction to handle.
+
+### Interaction logic
+
+#### `launch_interaction(session_index, session_info, candidate_receipt, candidate_hash)`
+
+1. Compute the threshold from the session info. It should be `f + 1`, where `n = 3f + k`, where `k in {1, 2, 3}`, and `n` is the number of validators.
+1. Set the various fields of `Interaction` based on the validator lists in `session_info`. Compute a random shuffling of the validator indices.
+1. Set the `to_state` sender to be equal to a clone of `state.from_interaction_tx`.
+1. Initialize `received_chunks` to an empty set, as well as `requesting_chunks`.
+
+Launch the interaction as a background task running `interaction_loop(interaction)`.
+
+#### `interaction_loop(interaction)`
+
+```rust
+// How many parallel requests to have going at once.
+const N_PARALLEL: usize = 50;
+```
+
+Loop:
+  * Poll for new updates from `requesting_chunks`. Check merkle proofs of any received chunks, and any failures should lead to issuance of a `FromInteraction::ReportPeer` message.
+  * If `received_chunks` has more than `threshold` entries, attempt to recover the data. If that fails, or a re-encoding of it doesn't match the expected erasure root, break and issue a `FromInteraction::Concluded(RecoveryError::Invalid)`. Otherwise, issue a `FromInteraction::Concluded(Ok(()))`.
```

Comment:
Recovery errors are a sign that some validators have issued faulty chunks right? If that is so shouldn't this issue be lifted to disputes? If that should what is the mechanism for proving to everyone that these validators issued these wrong chunks?

Reply:
Yeah, it should be escalated to disputes. My intention was to have approval voting handle that. The dispute participation subsystem will also make availability recovery requests and we'd expect it should encounter the same error. I elaborate on why that is below.

One of the things we are trying to avoid is the attack vector where some of the pieces successfully recover the data and some do not. Events might unfold in this order:
 * Validator A is given a set of pieces containing a bad piece and returns `RecoveryError::Invalid`. I am assuming that the adversary, by controlling the network and some bad validators, can ensure that only validator A can recover the bad piece.
 * Validators B-Z observe the dispute and each recovers a set of pieces which correctly recovery the `AvailableData`.
  * Validators B-Z vote to slash A, who has behaved according to protocol.

The way that we mitigate this attack is with by checking that "a re-encoding of it doesn't match the expected erasure root". Because we have this check, validators B-Z will all notice that the `erasure_root` in the candidate descriptor doesn't match, and thus return a `RecoveryError::Invalid`. Thus, the dispute will resolve against the candidate's backers.

cc @drahnr